Use `__slots__`

Currently we use this pattern to denote that `MyThing` will have a `prop_a` property.

```python
class MyThing:
    prop_a = None
```

It looks like if we instead use `__slots__` we could see 40-50% reduction in memory footprint.

```python
class MyThing:
    __slots__ = ["prop_a"]
```

source: 

- https://docs.python.org/3/reference/datamodel.html#slots
- http://book.pythontips.com/en/latest/__slots__magic.html

Interesting. So, it seems that if we would go down that path we'd lose out a lot of type safety features. By default `mypy` doesn't let us access attributes that aren't defined which is a good safety check to have. But it doesn't seem to recognize attributes defined via `slots` (leaving out for a moment that the string list donation doesn't let us specify the type of the property).

```python
class MyThing:
    __slots__ = ["prop_a"]

y = MyThing()
y.prop_a = 'test'
```

Leaves you with

```shell
$ mypy slots.py 
slots.py:5: error: "MyThing" has no attribute "prop_a"
```

Whereas the following works fine

```python
class MyThing:
    prop_a: str

y = MyThing()
y.prop_a = 'test'
```

**Plus**, it has the benefit of catching type bugs. E.g

```python
class MyThing:
    prop_a: str

y = MyThing()
y.prop_a = 4
```

```shell
$ mypy slots.py 
slots.py:14: error: Incompatible types in assignment (expression has type "int", variable has type "str")
```

>It looks like if we instead use __slots__ we could see 40-50% reduction in memory footprint.

I admit that I didn't really dive into *why* the `slots` construct would reduce the memory footprint that much but if it comes from the fact that we are pre-initializing all these attributes to `None` then I believe this won't be an issue with Python 3.6 anymore because there you don't need to assign `None` (or any other value) to an attribute just to denote it. You can simply declare it with its type and as far as I think this won't have any effect on runtime performance but be simply a hint for static analyzers such as `mypy`.

```python
class MyThing:
    prop_a: str
```

That's one downside of Python 3.5 that is also explicitly mentioned in [PEP526](https://www.python.org/dev/peps/pep-0526/).

>There's no way to annotate the type of an undefined variable; one needs to initialize it to None (e.g. a = None # type: int).

--- a/evm/vm/memory.py
+++ b/evm/vm/memory.py
@@ -17,7 +17,7 @@ class Memory(object):
     """
     VM Memory
     """
-    _bytes = None  # type: bytearray
+    __slots__ = ['_bytes']
     logger = logging.getLogger('evm.vm.memory.Memory')
 
     def __init__(self):

--- a/evm/vm/message.py
+++ b/evm/vm/message.py
@@ -17,20 +17,8 @@ class Message(object):
     """
     A message for VM computation.
     """
-    to = None
-    sender = None
-    value = None
-    data = None
-    gas = None  # type: int
-    depth = None
-
-    code = None
-    _code_address = None
-
-    create_address = None
-
-    should_transfer_value = None
-    is_static = None
+    __slots__ = ['to', 'sender', 'value', 'data', 'depth', 'gas', 'code', '_code_address',
+                'create_address', 'should_transfer_value', 'is_static', '_storage_address']
 
     logger = logging.getLogger('evm.vm.message.Message')
 
@@ -47,7 +35,7 @@ class Message(object):
                  should_transfer_value=True,
                  is_static=False):
         validate_uint256(gas, title="Message.gas")
-        self.gas = gas
+        self.gas = gas # type: int
 
         if to != CREATE_CONTRACT_ADDRESS:
             validate_canonical_address(to, title="Message.to")

--- a/evm/vm/message.py
+++ b/evm/vm/message.py
@@ -17,8 +17,10 @@ class Message(object):
     """
     A message for VM computation.
     """
-    __slots__ = ['to', 'sender', 'value', 'data', 'depth', 'gas', 'code', '_code_address',
-                'create_address', 'should_transfer_value', 'is_static', '_storage_address']
+    __slots__ = [
+        'to', 'sender', 'value', 'data', 'depth', 'gas', 'code', '_code_address',
+        'create_address', 'should_transfer_value', 'is_static', '_storage_address'
+    ]
 
     logger = logging.getLogger('evm.vm.message.Message')
 
@@ -35,7 +37,7 @@ class Message(object):
                  should_transfer_value=True,
                  is_static=False):
         validate_uint256(gas, title="Message.gas")
-        self.gas = gas # type: int
+        self.gas = gas  # type: int
 
         if to != CREATE_CONTRACT_ADDRESS:
             validate_canonical_address(to, title="Message.to")

--- a/evm/vm/stack.py
+++ b/evm/vm/stack.py
@@ -19,7 +19,7 @@ class Stack(object):
     """
     VM Stack
     """
-    values = None
+    __slots__ = ['values']
     logger = logging.getLogger('evm.vm.stack.Stack')
 
     def __init__(self):

--- a/evm/vm/stack.py
+++ b/evm/vm/stack.py
@@ -14,6 +14,9 @@ from evm.utils.numeric import (
     big_endian_to_int,
 )
 
+from eth_typing import Hash32
+from typing import List, Union
+
 
 class Stack(object):
     """
@@ -23,7 +26,7 @@ class Stack(object):
     logger = logging.getLogger('evm.vm.stack.Stack')
 
     def __init__(self):
-        self.values = []
+        self.values = []  # type: List[Union[int, Hash32]]
 
     def __len__(self):
         return len(self.values)

--- a/evm/vm/stack.py
+++ b/evm/vm/stack.py
@@ -14,8 +14,11 @@ from evm.utils.numeric import (
     big_endian_to_int,
 )
 
-from eth_typing import Hash32
-from typing import List, Union
+from typing import (  # noqa: F401
+    List,
+    Union
+)
+from eth_typing import Hash32  # noqa: F401
 
 
 class Stack(object):

--- a/evm/vm/state.py
+++ b/evm/vm/state.py
@@ -49,8 +49,7 @@ class BaseState(Configurable, metaclass=ABCMeta):
     #
     # Set from __init__
     #
-    _db = None
-    execution_context = None
+    __slots__ = ['_db', 'execution_context', 'account_db']
 
     computation_class = None  # type: Type[BaseComputation]
     transaction_context_class = None  # type: Type[BaseTransactionContext]

--- a/evm/vm/transaction_context.py
+++ b/evm/vm/transaction_context.py
@@ -11,9 +11,7 @@ class BaseTransactionContext:
     This immutable object houses information that remains constant for the entire context of the VM
     execution.
     """
-    _gas_price = None
-    _origin = None
-    _log_counter = None
+    __slots__ = ['_gas_price', '_origin', '_log_counter']
 
     def __init__(self, gas_price, origin):
         validate_uint256(gas_price, title="TransactionContext.gas_price")

